### ISSUE TITLE
LuaController serializer abuse.

### ISSUE THREAD

**Issue by beyondlimits** [minetest-mods/mesecons#415]
It is possible to **extremely** slow down server via forcing serializer to serialize a large fractal of nested tables, either in steps (my computer slows down after 24th):
```lua
interrupt(1)
mem.x = {mem.x, mem.x}
```
or bursting everything at once
```lua
for x = 1, 30 do
  o = {o, o}
end
mem.x = o
```

**Comment by numberZero** [comment]
@beyondlimits That particular bug can and should be fixed, but the core problem is unfixable. See https://github.com/minetest-mods/mesecons/issues/380#issuecomment-339689832 for some details; Lua just doesn’t offer enough sandboxing features.

**Comment by numberZero** [comment]
Possible solution: run the serializer under timeout.

**Comment by TheEt1234** [comment]
not allowing nested tables would solve this most likely

**Comment by fluxionary** [comment]
actually, [improvements in minetest's serializer](https://github.com/minetest/minetest/pull/11427) mean this isn't so easy to exploit currently. the following code gets up to 200 without any noticeable slowdown:

```lua
if event.type == "program" or event.type == "interrupt" then
 if pin.a then
  mem.i = (mem.i or 0) + 1
  digiline_send("LCD", tostring(mem.i))
  local o = mem.o or {}
  mem.o = {o, o}
 end

 interrupt(1)
end
```

trying to double the size of string instead of a table results in overheating after 16 iterations.

**Comment by TheEt1234** [comment]
oh cool, so it's basically fixed in the latest minetest version (?)

**Comment by fluxionary** [comment]
> oh cool, so it's basically fixed in the latest minetest version (?)

seems to be. possibly there's some way to exploit this, but i everything i can think of leads to overheating before major problems.

**Comment by Niklp09** [comment]
No, this is not fixed. It's still possible to create major lag spikes and I was even able to trigger an OOM server process kill on a 128GB RAM machine.
(I won't share the code here for obvious reasons)

**Comment by TheEt1234** [comment]
> not allowing nested tables would solve this most likely
> 
> edit: past me meant to say those `t = {} t.t = t` tables which i have no idea how to detect not allowing nested tables (as in `t = {} t.x = {}` would be a super breaking change)

How could this be fixed without it being a breaking change

**Comment by TheEt1234** [comment]
(ok i did not expect github to mention this. across those issues.. why does it do that, well the above is somewhat unrelated to the issue)
